src/main/java/uk/co/real_logic/benchmarks/rtt/Harness.java

```diff
@@ -0,0 +1,235 @@
+    }
+
+    private void doRun(
+        final int iterations,
+        final int messages,
+        final MessageProvider.Sender sender,
+        final MessageProvider.Receiver receiver,
+        final AtomicLong sentMessages,
+        final Histogram histogram)
+    {
+        CompletableFuture<?> receiverTask = runAsync(() -> receive(receiver, sentMessages, histogram));
+        sentMessages.set(send(iterations, messages, sender));
+        receiverTask.join();
+    }
+
+    void receive(final MessageProvider.Receiver receiver, final AtomicLong sentMessages, final Histogram histogram)
+    {
+        final NanoClock clock = this.clock;
+        final IdleStrategy idleStrategy = configuration.receiverIdleStrategy();
+
+        long sent = 0;
+        long received = 0;
+        while (true)
+        {
+            final long timestampNs = receiver.receive();
+            if (0 != timestampNs)
+            {
+                received++;
+                final long nowNs = clock.nanoTime();
+                final long durationNs = nowNs - timestampNs;
+                histogram.recordValue(durationNs);
+                idleStrategy.reset();
+            }
+            else
+            {
+                idleStrategy.idle();
+            }
+            if (0 == sent)
+            {
+                sent = sentMessages.get();
+            }
+            if (0 != sent && received == sent)
+            {
+                break;
+            }
+        }
+    }
+
+    long send(final int iterations, final int numberOfMessages, final MessageProvider.Sender sender)
+    {
+        final NanoClock clock = this.clock;
+        final PrintStream out = this.out;
+        final int burstSize = configuration.burstSize();
+        final int messageSize = configuration.messageSize();
+        final IdleStrategy idleStrategy = configuration.senderIdleStrategy();
+        final long intervalNs = NANOS_PER_SECOND * burstSize / numberOfMessages;
+        final long totalNumberOfMessages = (long)iterations * numberOfMessages;
+
+        final long startTimeNs = clock.nanoTime();
+        final long endTimeNs = startTimeNs + iterations * NANOS_PER_SECOND;
+        long sentMessages = 0;
+        long timestamp = startTimeNs;
+        long nowNs = startTimeNs;
+        long nextReportTimeNs = startTimeNs + NANOS_PER_SECOND;
+
+        while (true)
+        {
+            if (nowNs >= nextReportTimeNs)
+            {
+                final int elapsedSeconds = reportProgress(startTimeNs, nowNs, sentMessages);
+                nextReportTimeNs = startTimeNs + (elapsedSeconds + 1) * NANOS_PER_SECOND;
+            }
+
+            final int batchSize = (int)min(totalNumberOfMessages - sentMessages, burstSize);
+            int sent = sender.send(batchSize, messageSize, timestamp);
+            if (sent < batchSize)
+            {
+                idleStrategy.reset();
+                while ((sent += sender.send(batchSize - sent, messageSize, timestamp)) < batchSize)
```

Comment:
Should it not idle anytime we fail to send?

Reply:
You are right, I will change the loop to `do-while`, i.e. will `idle` before the next attempt to send.

Reply:
Done